src/sm.c

```diff
@@ -0,0 +1,331 @@
+        InterlockedIncrement64(&sm->e);
+        InterlockedExchange64(&sm->b_now, 0);
+    }
+}
+
+/*aims at getting a stable "e" for a stable "n" and returns the difference between them*/
+static LONG64 get_n_minus_e(SM_HANDLE sm)
+{
+    LONG64 e1, e2, n1, n2;
+    do
+    {
+        e1 = InterlockedAdd64(&sm->e, 0);
+        do
+        {
+            n1 = InterlockedAdd64(&sm->n, 0);
+            e2 = InterlockedAdd64(&sm->e, 0);
+            n2 = InterlockedAdd64(&sm->n, 0);
+        } while (n1 != n2);
+    } while (e1 != e2);
+
+    return n1 - e1;
+}
+
+/*sm_barrier_begin calls this. Also sm_close calls this, because _close is nothing else but another barrier*/
+static SM_RESULT sm_barrier_begin_internal(SM_HANDLE sm)
+{
+    SM_RESULT result;
+    LONG64 b_now = InterlockedOr64(&sm->b_now, 1);
+
+    /*Codes_SRS_SM_02_014: [ sm_close_begin shall set lowest bit of b_now to 1. ]*/
+    /*Codes_SRS_SM_02_028: [ If b_now has least significand bit set to 1 then sm_barrier_begin shall fail and SM_EXEC_REFUSED. ]*/
+    /*Codes_SRS_SM_02_020: [ If there was no sm_open_begin/sm_open_end called previously, sm_close_begin shall fail and SM_EXEC_REFUSED. ]*/
+    if ((b_now & 1) == 0)
+    {
+        InterlockedExchange(&sm->e_done, 0);
+
+        LONG64 n = InterlockedIncrement64(&sm->n);
+        (void)n;
+
+        /*Codes_SRS_SM_02_030: [ sm_barrier_begin shall succeed and return SM_EXEC_GRANTED. ]*/
+        /*Codes_SRS_SM_02_017: [ sm_close_begin shall succeed and return SM_EXEC_GRANTED. ]*/
+        result = SM_EXEC_GRANTED;
+
+        /*drain previous executions*/
+        /*Codes_SRS_SM_02_016: [ sm_close_begin shall wait for all previous operations to end. ]*/
+        /*Codes_SRS_SM_02_029: [ sm_barrier_begin shall wait for the completion of all the previous operations. ]*/
+        while (get_n_minus_e(sm) != 1)
+        {
+            if (InterlockedHL_WaitForValue(&sm->e_done, 1, INFINITE) != INTERLOCKED_HL_OK)
+            {
+                /*Codes_SRS_SM_02_031: [ If there are any failures then sm_barrier_begin shall fail and return SM_ERROR. ]*/
+                /*Codes_SRS_SM_02_034: [ If there are any failures then sm_close_begin shall fail and return SM_ERROR. ]*/
+                LogError("Catastrophic failure: InterlockedHL_WaitForValue(&sm->e_done, 1, INFINITE), name=%s, n=%" PRId64 "", sm->name, n);
+                InterlockedIncrement64(&sm->e); /*this is pretty fatal here - the wait failed... */
+                result = SM_ERROR;
+                break;
+            }
+            else
+            {
+                /*some thread thinks that n - 1 == e, the while condition will look at that throughly*/
+            }
+        }
+    }
+    else
+    {
+        /*Codes_SRS_SM_02_015: [ If setting the lowest bit b_now to 1 fails then sm_close_begin shall return SM_EXEC_REFUSED. ]*/
+        result = SM_EXEC_REFUSED;
+    }
+    return result;
+}
+
+
+SM_RESULT sm_close_begin(SM_HANDLE sm)
```

Comment:
>sm_close_begin [](start = 10, length = 14)

I know the spec says that user is in charge of pairing things. But!
We have a very limited number of actions that need pairing.
open, generic barrier (which includes close).

As the code is written now I could call begin_close and then sm_end, the code would execute but leave the state manager maybe in a bad state.
I would say that of course the caller code should not call sm_barrier_end. But it seems that state manager could detect these and at least log.
As a matter of fact state manager does detect calling begin_close before begin_open, but not between begin_open and end_open.

It seems to me that we have a pretty standard state machine, with several states:
NOT_OPENED, OPENING (between open_begin and open+end), OPENED, CLOSING (between close_begin and close_end).

And it looks to me that this state machine has some advantages:
- attempting to open something that opens already will fail (for example a user of the module attempting to call 2 times open)
- attempting to close something that is already closing will also fail
- attempting to close while in opening will also fail

Being that state manager does know of this state (it does have APIs with open/close) I believe it should validate the calls coming from the user. If we do not validate the calls for the user then the user module has to absolutely make sure that no call to begin_open happens between begin_open and end_open.
To achieve that the user module would have to have a state machine with NOT_OPENED, OPENING, OPEN, CLOSING.
We can push this check higher and higher on the stack until it finally will reach a user whose code is not under our control (SF, Hillary, etc.). And when that happens we need to handle these cases at that layer.

Then my question would be: if state manager knows about open/close, why not have the validation done in state manager?
Being a limited number of states I am sure they can be modeled with very few bits :-) and we have an int64 at our fingertips.
One such bit I believe is used today in b_now for storing that state manager is not open (-1) as opposed to another value (check made in open_begin).

I feel a bit bad to bring it up now after the code is written, but I think this will make things better for us and I really want to use this with minimal overhead for myself :-).
Maybe next time we should discuss it before the code is there... On the other hand I understand the need for a prototype first which probably let to the whole thing being written.

Reply:
As always, the anguish produced by a comment is directly proportional to the value of the comment.

Shall rethink.